alicloud/resource_alicloud_kvstore_instance.go

```diff
@@ -0,0 +1,309 @@
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"port": &schema.Schema{
+				Type:     schema.TypeInt,
+				Optional: true,
+				Computed: true,
+			},
+			"private_ip": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
+			"backup_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+
+func resourceAlicloudKVStoreInstanceCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	conn := client.rkvconn
+
+	request, err := buildKVStoreCreateRequest(d, meta)
+	if err != nil {
+		return err
+	}
+
+	resp, err := conn.CreateInstance(request)
+
+	if err != nil {
+		return fmt.Errorf("Error creating Alicloud db instance: %#v", err)
+	}
+
+	d.SetId(resp.InstanceId)
+
+	// wait instance status change from Creating to Normal
+
+	if err := client.WaitForRKVInstance(d.Id(), Normal, DefaultLongTimeout); err != nil {
+		return fmt.Errorf("WaitForInstance %s got error: %#v", Running, err)
+	}
+
+	return resourceAlicloudKVStoreInstanceRead(d, meta)
+}
+
+func resourceAlicloudKVStoreInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	conn := client.rkvconn
+	d.Partial(true)
+
+	if d.HasChange("instance_class") {
+		request := r_kvstore.CreateModifyInstanceSpecRequest()
+		request.InstanceId = d.Id()
+		request.InstanceClass = d.Get("instance_class").(string)
+		if _, err := conn.ModifyInstanceSpec(request); err != nil {
+			return err
+		}
+		// wait instance status is Normal after modifying
+		if err := client.WaitForRKVInstance(d.Id(), Normal, DefaultLongTimeout); err != nil {
+			return fmt.Errorf("WaitForInstance %s got error: %#v", Running, err)
+		}
+
+		d.SetPartial("instance_class")
+	}
+
+	if d.HasChange("instance_name") {
+		request := r_kvstore.CreateModifyInstanceAttributeRequest()
+		request.InstanceId = d.Id()
+		request.InstanceName = d.Get("instance_name").(string)
+		// wait instance status is Normal before modifying
+		if err := client.WaitForRKVInstance(d.Id(), Normal, DefaultLongTimeout); err != nil {
+			return fmt.Errorf("WaitForInstance %s got error: %#v", Running, err)
+		}
+		if _, err := conn.ModifyInstanceAttribute(request); err != nil {
+			return fmt.Errorf("ModifyRKVInstanceDescription got an error: %#v", err)
+		}
+		// wait instance status is Normal after modifying
+		if err := client.WaitForRKVInstance(d.Id(), Normal, DefaultLongTimeout); err != nil {
+			return fmt.Errorf("WaitForInstance %s got error: %#v", Running, err)
+		}
+
+		d.SetPartial("instance_name")
+	}
+
+	d.Partial(false)
+	return resourceAlicloudKVStoreInstanceRead(d, meta)
+}
+
+func resourceAlicloudKVStoreInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*AliyunClient)
+	instance, err := client.DescribeRKVInstanceById(d.Id())
+	if err != nil {
+		if IsExceptedError(err, InvalidKVStoreInstanceIdNotFound) {
+			d.SetId("")
```

Comment:
I think line 195 can be moved into method DescribeRKVInstanceById and modify line 195 as `if NotFoundError(err)`.

Reply:
Not sure what you mean here: line 195 cannot be replaced by `NotFoundError(err)` since this method is checking for the error code `Forbidden.InstanceNotFound` which is different from kvstore.
Can you provide a complete sample? Maybe I'm misunderstanding you.

Reply:
done